temporian/implementation/numpy/operators/add_index.py

```diff
@@ -32,24 +33,20 @@ def __call__(self, input: EventSet) -> Dict[str, EventSet]:
 
         dst_data = {}
         for src_index, src_data in input.data.items():
-            # Maps, for each new index key, the indices of the events in
-            # src_data.
-            #
-            # TODO: Do more efficiently. E.g. with numpy masks.
+            index_features = [src_data.features[i] for i in new_index_idxs]
+            (
+                group_keys,
+                row_idxs,
+                group_begin_idx,
+            ) = operators_cc.add_index_compute_index(index_features)
 
-            new_index_to_value_idxs = defaultdict(list)
-            # TODO: This is slow. Speed-up.
-            for event_idx, new_index in enumerate(
-                zip(*[src_data.features[f_idx] for f_idx in new_index_idxs])
-            ):
-                new_index = tuple(new_index)
-                new_index_to_value_idxs[new_index].append(event_idx)
-
-            for new_index, example_idxs in new_index_to_value_idxs.items():
-                # Note: The new indexes added after the existing ones.
-                dst_index = src_index + new_index
+            for group_idx, group_key in enumerate(group_keys):
+                dst_index = src_index + group_key
                 assert isinstance(dst_index, tuple)
 
+                example_idxs = row_idxs[
+                    group_begin_idx[group_idx] : group_begin_idx[group_idx + 1]
+                ]
```

Comment:
The use of `index` (or `idx`) in this op is very confusing, as it refers to the temporian concept of index and also to array positions.
In the future we should consider alternative names for ordinal positions, like `pos`, `loc`, `ord`, `num`.
I think that `index` is better in any other context, but here it's ambiguous and also leads to these weird names like "example", I guess to avoid `index_idxs` (?)

Reply:
let's figure a way next meeting :).

So far, we are using "index" to talk about Temporian index, and "idx" to reference integers that define a position in a list/vector.